src/main/python/tests/examples/tutorials/test_adult.py

```diff
@@ -387,6 +387,11 @@ def test_level2(self):
         ################################################################################################################
         X1, M1 = X1.transform_encode(spec=jspec)
 
+        # better alternative for encoding
+        # X1, M = F1.transform_encode(spec=jspec)
+        # X2 = F2.transform_apply(spec=jspec, meta=M)
+        # testX2 = X2.compute(True)
```

Comment:
@Baunsgaard
Hi, we tried to implement a better version of the encoding using transform_apply, because otherwise we would calculate statistics for imputation on the whole data instead of just the training data.
Unfortunately we ran into the problem that the labels in the train data are slightly different than the labels in the test data ("<=50K" != "<=50K."), which hinders us in using the encoding M for the test data. We tried different methods for correcting the labels in the test data before encoding, however the main problem is that we have not found a good way to apply changes in a column of a frame, and we are also unable to use frames as arguments for a dml script function (it seems to be not supported yet with python?).
The simplest solution would be to remove the "." at the end of the test labels in the file itself. Other possible workarounds using just systemds functions seem to be quite complex and would probably miss the main goal of this tutorial.
How should we proceed?

Reply:
two things here.

1. Yes, if there is a different number of labels in the training vs the test data, then the transform_apply does not work, currently there is no way around this. Do you really have 50k different classes in one of the features? I think you might be using a wrong encoding scheme for some of the columns.
2. Frames should be supported, but they are very new so there are bound to be bugs, the function definitions should specify frame if the input type is frame otherwise you should not call a function with frames, could you tell me which function you are trying to call, then i can try to fix it if there is a bug?

Reply:
1. There are only 2 different labels to predict, if a person makes "<=50K" a year or ">50K" a year. The 2 labels in the training set are denoted correctly as  either "<=50K" and ">50K", however in the test set they are called "<=50K."  and ">50K.". The dot at the end of the test set labels hinders us to use the transform_apply function correctly.
2. The dml function we want to use should be able to replace a string in the target column of the frame, similar to 

```
replace_target_frame = function(String replacement, String to_replace, Frame[Unknown] X)
  return(Frame[Unknown] X)
{
  for (i in 1:nrow(X)) {
    if (as.scalar(X[i, ncol(X)]) == to_replace) {
      X[i, ncol(X)] = replacement;
    }
  }
}
```

However when trying to load the function into python, we get the error 
> "NotImplementedError: Not Implemented type parsing for function def: Frame[Unknown]X"

Reply:
1. Okay now i understand the problem... a classic ... "someone made an error when making the dataset"...
2. Since you have this issue i just extended the frame to support the replace operation simply use

replace(target=X, pattern="<=50K.", replacement="<=50K")

Reply:
Thank you very much! Still I am confused how to use the replace function in the python environment? Because we load the data in python into a frame, and when trying to use `sds.source` with a dml file that includes a function with a frame as an argument, it throws the above error.

Reply:
well i guess, i did not add it to the python API... will do.

Reply:
Should be there now ... if you have a matrix or a frame simply call .replace("target","pattern")

Reply:
Thank you!
One problem still remains: The replace works for ">50K.", however it does not for "<=50K.". I played around a bit, and my guess is that the "<=" is interpreted as some kind of operator, because when I execute
`F2 = F2.replace("\<=50K.", "<=50K")`
I get this error

> Exception       : An error occurred while calling o0.prepareScript.
: org.apache.sysds.parser.ParseException: 
The following 4 parse issues were encountered:
#1 [line 2:32] [Syntax error] -> V1=replace(target=V0,pattern='\<=50K.',replacement='<=50K');
   no viable alternative at input 'V1=replace(target=V0,pattern==' ([@25,226:226='=',<7>,2:32])
#2 [line 2:32] [Syntax error] -> V1=replace(target=V0,pattern='\<=50K.',replacement='<=50K');
   extraneous input '=' expecting {'(', '[', '-', '+', '!', 'TRUE', 'FALSE', ID, INT, DOUBLE, COMMANDLINE_NAMED_ID, COMMANDLINE_POSITION_ID, STRING} ([@25,226:226='=',<7>,2:32])
#3 [line 2:32] [Validation error] -> V1=replace(target=V0,pattern='\<=50K.',replacement='<=50K');
   cannot parse the int value: '50K'
#4 [line 2:35] [Syntax error] -> V1=replace(target=V0,pattern='\<=50K.',replacement='<=50K');
   extraneous input 'K' expecting {')', ',', '^', '-', '+', '%*%', '%/%', '%%', '*', '/', '>', '>=', '<', '<=', '==', '!=', '&', '&&', '|', '||'} ([@27,229:229='K',<60>,2:35])